pickfirst.go

```diff
@@ -50,23 +50,37 @@ type pickfirstBalancer struct {
 }
 
 func (b *pickfirstBalancer) ResolverError(err error) {
-	switch b.state {
-	case connectivity.TransientFailure, connectivity.Idle, connectivity.Connecting:
-		// Set a failing picker if we don't have a good picker.
-		b.cc.UpdateState(balancer.State{ConnectivityState: connectivity.TransientFailure,
-			Picker: &picker{err: fmt.Errorf("name resolver error: %v", err)},
-		})
-	}
 	if logger.V(2) {
 		logger.Infof("pickfirstBalancer: ResolverError called with error %v", err)
 	}
+	if b.sc == nil {
+		b.state = connectivity.TransientFailure
+	}
+
+	if b.state != connectivity.TransientFailure {
+		// The picker will not change since the balancer does not currently
+		// report an error.
+		return
+	}
+	b.cc.UpdateState(balancer.State{
+		ConnectivityState: connectivity.TransientFailure,
+		Picker:            &picker{err: fmt.Errorf("name resolver error: %v", err)},
+	})
 }
 
 func (b *pickfirstBalancer) UpdateClientConnState(cs balancer.ClientConnState) error {
 	if len(cs.ResolverState.Addresses) == 0 {
+		if b.sc != nil {
+			// If the resolver returned empty addresses after it gave us non-empty
+			// addresses, we need to close the existing subConn. Setting sc to nil
+			// here will cause us to return an error picker from ResolverError().
```

Comment:
Should we not update our picker?  I guess we wait for the subconn to be shutdown and then update our state?  But in UpdateSCS it just sets `b.sc` to `nil` when we get a shutdown, and we don't update our picker.  I think something is missing.

Reply:
> Should we not update our picker?

Hmm ... are you suggesting that we not update the picker here or asking why we are not updating the picker? If it is the latter, then here is my answer:
After setting `b.subConn` to `nil` here, we end up calling `b.ResolverError` below, and this will return an error picker because it sees `b.subConn == nil`. I modeled this based on the implementation in `round_robin`.

Reply:
Oh yeah, the latter is what I missed, and I missed this interaction for some reason.  Thanks.